src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,160 @@
+    internal sealed class CSharpForeachToForCodeRefactoringProvider : AbstractForeachToForCodeRefactoringProvider
+    {
+        public CSharpForeachToForCodeRefactoringProvider()
+        {
+        }
+
+        protected override SyntaxNode GetForeachStatement(SyntaxToken token)
+        {
+            var foreachStatement = token.Parent.FirstAncestorOrSelf<ForEachStatementSyntax>();
+            if (foreachStatement == null)
+            {
+                return null;
+            }
+
+            // support refactoring only if caret is in between "foreach" and ")"
+            var scope = TextSpan.FromBounds(foreachStatement.ForEachKeyword.Span.Start, foreachStatement.CloseParenToken.Span.End);
+            if (!scope.IntersectsWith(token.Span))
+            {
+                return null;
+            }
+
+            // check whether there is any comments between foreach and ) tokens
+            // if they do, we don't support conversion.
+            foreach (var trivia in foreachStatement.DescendantTrivia(n => n == foreachStatement || scope.Contains(n.FullSpan)))
+            {
+                if (trivia.Span.End <= scope.Start ||
+                    scope.End <= trivia.Span.Start)
+                {
+                    continue;
+                }
+
+                if (trivia.Kind() != SyntaxKind.WhitespaceTrivia &&
+                    trivia.Kind() != SyntaxKind.EndOfLineTrivia)
+                {
+                    // we don't know what to do with these comments
+                    return null;
+                }
+            }
+
+            return foreachStatement;
+        }
+
+        protected override (SyntaxNode start, SyntaxNode end) GetForeachBody(SyntaxNode node)
+        {
+            var foreachStatement = (ForEachStatementSyntax)node;
+            return (foreachStatement.Statement, foreachStatement.Statement);
+        }
+
+        protected override void ConvertToForStatement(SemanticModel model, ForeachInfo foreachInfo, SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var parseOption = model.SyntaxTree.Options;
+            var foreachStatement = (ForEachStatementSyntax)foreachInfo.ForeachStatement;
+
+            // first, see whether we need to introduce new statement to capture collection
+            var foreachCollectionExpression = GetForeachCollection(foreachStatement);
+            var collectionVariableName = foreachCollectionExpression.ToString();
+            if (foreachInfo.RequireCollectionStatement)
+            {
+                collectionVariableName = CreateUniqueMethodName(model, foreachStatement, "list");
+
+                var collectionStatement = SyntaxFactory.LocalDeclarationStatement(
+                    SyntaxFactory.VariableDeclaration(
+                        SyntaxFactory.IdentifierName("var"),
```

Comment:
Should use var/type-name depending on the user's preference.

Note: consider just using SyntaxGenerator to simplify this a lot.

Reply:
I looked SyntaxGenerator.LocalDeclaration and it didn't use Options so that seems still not pick "var" or user types based on options. for now, I didn't do "var" or explicit type thing since I couldn't find one central stuff that does it for me. will look UseExplicitType refactoring rather to see what it does.

Reply:
I will use the TypeStyleHelper stuff you added in your PR once that is checked in